src/compiler/checker.ts

```diff
@@ -34550,6 +34550,9 @@ namespace ts {
             }
 
             const location = isBinaryExpression(condExpr) ? condExpr.right : condExpr;
+            if (isBinaryExpression(condExpr) && condExpr.operatorToken.kind === SyntaxKind.BarBarToken) {
+                checkTestingKnownTruthyCallableOrAwaitableType(condExpr.left, type, body);

```

Comment:
I don't think recursion is a good idea here, because binary expressions might be pretty long, and because the core of `checkTestingKnownTruthyCallableOrAwaitableType` is not too complex.

I'd move the `getSignatureOfType` check before `location`'s declaration, then put the rest of the function in a helper that's called in a loop, something like:

```ts
            helper(condExpr, body)
            while (isBinaryExpression(condExpr) && condExpr.operatorToken.kind === SyntaxKind.BarBarToken) {
                condExpr = condExpr.left
                helper(condExpr, body)
            }
```

(although you might be able to write it more elegantly)